### PR TITLE
introduce schedulerOptions to optimize scheduler instantiation

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
1. Currently  when we instantce `NewScheduler`, we passed in a named `Options` struct which contains too much useless fields, this PR made the parameters passed in when needed.  It's also easy to mock scheduler.
2. Add scheduler unit tests.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

### CODE REVIEW

**Comment by kerthcet** [comment]
cc @huone1

**Comment by RainbowMango** [comment]
Is this part of #1371 ?

File: pkg/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			sche, err := NewScheduler(dynamicClient, karmadaClient, kubeClient, tc.opts...)
+			if err != nil {
+				t.Errorf("create scheduler error: %s", err)
+			}
+			if diff := cmp.Diff(tc.enableSchedulerEstimator, sche.enableSchedulerEstimator); diff != "" {
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler_test.go:44]: Is it the right place to introduce `cmp` here? 
I tried to hack this test and make it fail. It doesn't looks better:
```
=== RUN   TestCreateScheduler/scheduler_with_enableSchedulerEstimator_enabled
    scheduler_test.go:45: unexpected enableSchedulerEstimator (-want, +got):
          bool(
        -       false,
        +       true,
          )
```

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -105,30 +142,31 @@ func NewScheduler(dynamicClient dynamic.Interface, karmadaClient karmadaclientse
...
-		bindingLister:            bindingLister,
-		policyLister:             policyLister,
-		clusterBindingLister:     clusterBindingLister,
-		clusterPolicyLister:      clusterPolicyLister,
-		clusterLister:            clusterLister,
-		informerFactory:          factory,
-		queue:                    queue,
-		Algorithm:                algorithm,
-		schedulerCache:           schedulerCache,
-		enableSchedulerEstimator: opts.EnableSchedulerEstimator,
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:120]: nice finding.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -86,8 +85,41 @@ type Scheduler struct {
...
+	// enableSchedulerEstimator represents whether the accurate scheduler estimator should be enabled.
+	enableSchedulerEstimator bool
+	// schedulerEstimatorTimeout specifies the timeout period of calling the accurate scheduler estimator service.
+	schedulerEstimatorTimeout metav1.Duration
+	// schedulerEstimatorPort is the port that the accurate scheduler estimator server serves at.
+	schedulerEstimatorPort int
+}
+
+// Option configures a Scheduler
+type Option func(*schedulerOptions)
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:98]: Would it be a little bit confusing for this Option `function` and [struct](https://github.com/karmada-io/karmada/blob/28a6c2c7672a5efd504c9c264bf2c6f53e5e89b4/cmd/scheduler/app/options/options.go#L28)?

**Review by RainbowMango — COMMENTED** [review]
/assign @lonelyCZ @huone1

**Comment by kerthcet** [comment]
> Is this part of #1371 ?

Somehow related.

File: pkg/scheduler/scheduler_test.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go:44]: got it.

File: pkg/scheduler/scheduler.go (hunk 3 above)
Comment (kerthcet) [reply pkg/scheduler/scheduler.go:98]: They belong to different packages, looks good to me. Or change to `type Configurator func(*schedulerConfigurators)` ?

**Comment by huone1** [comment]
it is similar to the k8s framework，and  is the ""cmp""  necessary ?

File: pkg/scheduler/scheduler.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/scheduler/scheduler.go:98]: I'm thinking about why we need to introduce the `Option`, can we reuse the legacy option? https://github.com/karmada-io/karmada/blob/28a6c2c7672a5efd504c9c264bf2c6f53e5e89b4/cmd/scheduler/app/options/options.go#L28

I know, there is a lot of `option`s in the Kubernetes scheduler, but my concern is the readability.

Comment (huone1) [reply pkg/scheduler/scheduler.go:98]: it is better to describe the benefits to import the **Option**

Comment (kerthcet) [reply pkg/scheduler/scheduler.go:98]: Yes, it's not necessary but I think meaningful. [karmada/cmd/scheduler/app/options/options.go](https://github.com/karmada-io/karmada/blob/28a6c2c7672a5efd504c9c264bf2c6f53e5e89b4/cmd/scheduler/app/options/options.go#L28) holds lots of options we don't need to instantiate scheduler, as time goes on, there will be more options introduced. It's hard to maintain as we have to check the code to see which options we need and which one is required by other components, so we should only keep the necessary ones. Also it's easy to mock the `NewScheduler` tests then.

**Comment by kerthcet** [comment]
> it is similar to the k8s framework，and is the ""cmp"" necessary ?

Well, it's also not necessary, but I prefer to introduce this lib to our tests as we don't have to think about the proper words like `We expected xxxx, but got yyy', we can unified the test like 
```
			if diff := cmp.Diff(tc.want, got); diff != "" {
				t.Errorf("Unexpected xxx (-want,+got):\n%s", diff)
			}
```
It also reduces the burden of reviewers.  But I'm ok with introduce this lib or not, in the pr or that pr.  WDYT?

**Comment by kerthcet** [comment]
I'll revert the `cmp.Diff` change then.

**Comment by huone1** [comment]
> I'll revert the `cmp.Diff` change then.

@RainbowMango  what do you think about the lib **cmp** ?

**Comment by kerthcet** [comment]
> > I'll revert the `cmp.Diff` change then.
> 
> @RainbowMango what do you think about the lib **cmp** ?

Removed the cmp package.

File: pkg/scheduler/scheduler_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}{
+		{
+			name:                     "scheduler with default configuration",
+			opts:                     nil,
+			enableSchedulerEstimator: false,
+		},
+		{
+			name: "scheduler with enableSchedulerEstimator enabled",
+			opts: []Option{
+				WithEnableSchedulerEstimator(true),
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler_test.go:31]: How about the `WithSchedulerEstimatorTimeout` and `WithSchedulerEstimatorPort` ? should we test them as well?

Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go:31]: I'll add tests for `WithSchedulerEstimatorPort`, but `Timeout` is hard to test unless we add a getter method. Considering `WithSchedulerEstimatorPort` somehow behaves the same as `WithSchedulerEstimatorTimeout`, I think it's ok.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**